File: packages/guards/src/input-length-anomaly.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface InputLengthAnomalyOptions {
  action: 'block' | 'warn';
  minLength?: number;
  maxLength?: number;
}

export function inputLengthAnomaly(options: InputLengthAnomalyOptions): Guard {
  const minLen = options.minLength ?? 3;
  const maxLen = options.maxLength ?? 50000;

  return {
    name: 'input-length-anomaly',
    version: '0.1.0',
    description: 'Detects anomalous input lengths that might indicate attacks',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];
      const len = text.length;

      if (len < minLen) issues.push('too-short');
      if (len > maxLen) issues.push('too-long');

      const triggered = issues.length > 0;

      return {
        guardName: 'input-length-anomaly',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { issues, length: len, minLength: minLen, maxLength: maxLen }
          : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/input_length_anomaly.py
"""Input length anomaly detection guard."""

import time

from open_guardrail.core import GuardResult


class _InputLengthAnomaly:
    def __init__(
        self,
        *,
        action: str = "block",
        min_length: int = 3,
        max_length: int = 50000,
    ) -> None:
        self.name = "input-length-anomaly"
        self.action = action
        self._min = min_length
        self._max = max_length

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []
        length = len(text)

        if length < self._min:
            issues.append("too-short")
        if length > self._max:
            issues.append("too-long")

        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="input-length-anomaly",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                f"Input length anomaly: {', '.join(issues)}"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "issues": issues,
                    "length": length,
                    "min_length": self._min,
                    "max_length": self._max,
                }
                if triggered
                else None
            ),
        )


def input_length_anomaly(
    *,
    action: str = "block",
    min_length: int = 3,
    max_length: int = 50000,
) -> _InputLengthAnomaly:
    return _InputLengthAnomaly(
        action=action,
        min_length=min_length,
        max_length=max_length,
    )
